File: swift/Solutions/Day04.swift
//
//  Day04.swift
//  advent_of_code_2023
//
//  Created by Kushantha Attanayake on 12/3/23.
//

import Foundation

struct Day04: Day{
    func part1Solution(input: [String]) -> String {
        var answer = 0
        input.forEach { game in
            let numbers = game.split(separator: ":")[1].split(separator:"|")
            let winningNumbers = numbers[0].split(separator:" ").map { Int($0)!}
            let actualNumbers = numbers[1].split(separator:" ").map { Int($0)!}
            let intersection = Set(winningNumbers).intersection(actualNumbers)
            if(!intersection.isEmpty){
                answer += Int(pow(Double(2), Double(intersection.count-1)))
            }
        }
        return "\(answer)"
    }
    
    func part2Solution(input: [String]) -> String {
        var lineMultipliers = [Int](repeating: 1, count: input.count)
        for (idx, game) in input.enumerated(){
            let numbers = game.split(separator: ":")[1].split(separator:"|")
            let winningNumbers = numbers[0].split(separator:" ").map { Int($0)!}
            let actualNumbers = numbers[1].split(separator:" ").map { Int($0)!}
            
            let intersection = Set(winningNumbers).intersection(actualNumbers)
            
            let multiplier = lineMultipliers[idx]
            if(!intersection.isEmpty){
                for gameNum in 1...intersection.count {
                    lineMultipliers[idx+gameNum] += multiplier
                }
            }
        }
        
        return "\(lineMultipliers.reduce(0, +))"
    }
    
    
}

File: kotlin/src/main/kotlin/solutions/Day04.kt
package solutions

import java.lang.Math.pow

class Day04: Day {
    override fun part1Solution(input: List<String>): String {
        var answer = 0
        input.forEach { game ->
            val (winning, regular) = game.split(":")[1].split("|").map {nums ->
                nums.split(" ").filter { it.isNotEmpty() }.map { it.toInt() }
            }
            winning.intersect(regular.toSet()).let {
                if (it.isNotEmpty()) answer += pow(2.0, (it.size-1).toDouble()).toInt()
            }
        }
        return answer.toString()
    }

    override fun part2Solution(input: List<String>): String {
        val multipliers = IntArray(input.size){ 1 }
        input.forEachIndexed { index, game ->
            val (winning, regular) = game.split(":")[1].split("|").map { nums ->
                nums.split(" ").filter { it.isNotEmpty() }.map { it.toInt() }
            }
            for(gameNum in 1..winning.intersect(regular.toSet()).size){
                multipliers[index+gameNum] += multipliers[index]
            }
        }
        return multipliers.sum().toString()
    }
}